### ISSUE TITLE
Source2

### ISSUE THREAD

**Issue by ykim** [skadistats/clarity#42]
I'm not too sure how feasible it is going to be for clarity to integrate Source2 with regards to Source1, but I figure it's best to make an issue for this to put it on the roadmap at least. For now, I believe alice2 is capable of parsing a Source2-generated replay file. This could be used as a starting point.

**Comment by spheenik** [comment]
it's issue #42 :D

**Comment by ykim** [comment]
According to [this](http://dev.dota2.com/showthread.php?t=107817&page=17208&p=1234472&viewfull=1#post1234472), it's looking like Source2 might be coming out in conjunction with Source for backwards compatibility. I can start looking into the Source2 format.

**Comment by howardc93** [comment]
https://github.com/SteamDatabase/GameTracking/tree/master/Protobufs/dota_s2 has protobufs for source2.  

Will there need to a be a significant rewrite of the decoders, etc., or can the new protobufs just be dropped in?

**Comment by spheenik** [comment]
There is 2 types of replays: "console recorded", and "provided as a complete demo". The console recorded ones are basically a dump of the network traffic, and contain some optimizations and other unique features (for example using 2 switchable baselines). See Issue #36, I still have problems parsing S1 console replays. 

The replays from S2 we have seen so far have all been console recorded too, and they contain a whole lot of new optimizations and other changes, some of which have been decyphered, while others haven't. So this is all for S2-console-recorded:

From what I know, the status quo now is that it's possible to unpack S2-versions of CDemoPacket, but impossible to correctly read string tables. No one has tried reading entities yet, and if they got equal treatment as the StringTables, it'll certainly not be a cakewalk.

As for S2-normal-replays: Volvo plz send me one!
And to speculate: S2-normal might be closer to S1-normal than expected. Also, I expect the two clients to run in parallel for a while, and the S2 one to be able to play old replays. 

TL;DR:  Always in motion is the future... ;)

**Comment by howardc93** [comment]
Possible source2-normal replay:

https://drive.google.com/file/d/0B37dmavKxjUmbkY3ZC01WEpEeWM/view?usp=sharing

**Comment by howardc93** [comment]
Is the plan to have clarity support parsing both replay types (by reading the header), or will the source 2-capable clarity be its own project?

**Comment by spheenik** [comment]
The plan is to create a single clarity-protobufs jar containing the bufs for both versions in separate packages, and to abstract away differences between both engine versions (praying every day that it'll be possible)

**Comment by spheenik** [comment]
btw. how do you know the replay is a normal one?

**Comment by howardc93** [comment]
I don't exactly (haven't tried to parse it yet), but I believe it's what
Dotabuff has been using to develop their Source 2 parser.  Dotabuff
announced on Twitter a few days ago that they had encountered a Source 2
replay "in the wild", and I suspect this is it (although I can't be sure).

**Comment by ykim** [comment]
@howardc93 : How did you retrieve this match? Is it the same one that Dotabuff linked to on Twitter?

**Comment by howardc93** [comment]
It was linked to in their Source 2 parser repo.  I don't know for sure if it is the same replay.

**Comment by ykim** [comment]
@spheenik : From looking at the parser that dotabuff has been working on, the structure of the replay is somewhat similar to S1. The differences I'm noting are as follows:

- Magic is changed. This makes it easy to load different protobuf parsers.
- Replay size bits are 8 bytes instead of 4 bytes.
- Additional protobufs are added, such as spawn groups.
- Some data structures are structured differently, such as sendtables.

From that, I'm guessing that your approach to supporting both should work.

**Comment by albertcui** [comment]
Hey Martin! What's the roadmap looking for Source 2? Valve is looking to release Source 2 in the next few days, so @howardchung and I are worried about what to do for YASP. What's left to be done? Wondering if we can help finish.

Thanks!

**Comment by spheenik** [comment]
This is all that’s left to implement:

https://github.com/skadistats/clarity/blob/master/src/main/java/skadistats/clarity/model/s2/S2DTClass.java#L42-L50

Rest is working perfectly (until todays update, which I could not check yet).
Do you have any info on the release date?
I was planning to finish it on the weekend...

**Comment by albertcui** [comment]
Apparently it's in the [next day or two](http://blog.dota2.com/2015/09/introducing-dota-levels/).

**Comment by spheenik** [comment]
ouch!
I’ll hack it together tonight…
Gimme an email address (quick, next 5 minutes) and I’ll send you the extract of the send tables.
That way, you can see where the properties you need are at in S2…

**Comment by spheenik** [comment]
http://schrodt.org/baseline/flattables.txt.zip

**Comment by spheenik** [comment]
Additionaly, here is an S2 baseline for [CDOTA_PlayerResource](http://schrodt.org/baseline/CDOTA_PlayerResource.txt). You got to download the file and open in a text editor, since it's too wide to view in a browser .(

**Comment by spheenik** [comment]
@albertcui:

It's done. I hacked together a functional version. It's still slow, and might contain bugs, but at least the matchend example is working and should get you going:

https://github.com/skadistats/clarity-examples/blob/master/src/main/java/skadistats/clarity/examples/matchend/Main.java#L37-L81

gn8 :)

**Comment by spheenik** [comment]
and forget about those baselines I sent this morning.
Just run the serializers example on a replay, it'll spit out the flattables.txt, and run the test example, it'll spit out tables of all baselines.

**Comment by aybassiouny** [comment]
Congrats!

I wanna help and I have some free time - any suggestions where to start?

**Comment by albertcui** [comment]
Thanks for the quick work! We're up. Also, thanks for the example and the sendtables. Super useful :)

**Comment by spheenik** [comment]
@aybassiouny 

> I wanna help and I have some free time - any suggestions where to start?

Depends on how deep your knowledge on the replay data is.
Several things need come to my mind:

* the combat log example needs to have new cases added (they do not emit proper strings for new events)
* A test infrastructure has to be implemented (not something testing the obvious, but rather a generator for fixtures and a matching java class that tests those)
* Optimization of the S2 parser

**Comment by enche** [comment]
@spheenik 
> Depends on how deep your knowledge on the replay data is. [...]

I also want to help & have free time! Do you have a suggestion of where to start learning about the replay data? Of those three bullets, I'd want to go for optimizing the S2 parser. 

Side note: the combat log in game is very buggy, e.g.: Faceless Void is named "f", others are blank; I don't know if this means anything other than a front end engine issue on Valve's end though. I'm unfamiliar with S1 logs too, to compare; as well as combat log updates of the past by Valve.

**Comment by spheenik** [comment]
Ok, sure, but for optimizing it, you first have to understand how things are done atm. Start by understanding S2FieldReader.java, and how it's working together with Serializer.java and different implementations of Field.java.

I just tried combat log (match id 1785874713): works fine (apart from those missing entries). 
But you mean in the game... Hmm.

**Comment by spheenik** [comment]
Also, there's a new example since today: dumpbaselines. 
You can use it to dump the instancebaselines of all entities to get an easy look at how they are structured.

**Comment by spheenik** [comment]
Done with 2.1-SNAPSHOT
